Doc: Revise section on info_plist for plistlib functionality.

itsayellow opened the pull request:
See Issue #3540 
As plistlib has been added to the code to process the info_plist build argument dict, info_plist can now accept nested python objects, and python datatypes other than strings.  This update to the documentation reflects these new capabilities and includes new examples, as well as removing descriptions of limitations that no longer exist.

htgoebel commented on doc/spec-files.rst:
```diff
@@ -383,36 +383,47 @@ also work in retina screen::
+XML.)  The following python data types are translated to the proper
+:file:`Info.plist` types: strings, integers, floats, booleans, tuples, lists,
+dictionaries (but only with string keys), Data, bytes, bytesarray, and
+datetime.datetime objects.
```
IMO it would be okay to only refer to `plistlib` for details.

htgoebel commented on doc/spec-files.rst:
```diff
@@ -383,36 +383,47 @@ also work in retina screen::
-and follow it with a statement such as::
-
-    cp -f Info.plist dist/myscript.app/Contents/Info.plist
+    app = BUNDLE(exe,
```
I suggest having only one example, which I asumme to be enough. So the order would be something like 
- old lines up to 373 ("...parameter to the BUNDLE call")
- our new lines 390-393
- this large example
- short explanation (based on old lines 378-379 and your new lines 398-407)

htgoebel commented on doc/spec-files.rst:
```diff
@@ -383,36 +383,47 @@ also work in retina screen::
+                        'LSItemContentTypes': [
+                            'com.example.myformat'
+                            ]
+                        'LSHandlerRank': 'Owner'
```
Trailing commas (separators) are missing at most of the lines above.

htgoebel commented on doc/spec-files.rst:
```diff
@@ -383,36 +383,47 @@ also work in retina screen::
+                        'CFBundleTypeIconFile': 'MyFileIcon.icns'
+                        'LSItemContentTypes': [
+                            'com.example.myformat'
+                            ]
```
Use only one line for `'LSItemContentTypes': ['com.example.myformat'],`?

htgoebel commented on doc/spec-files.rst:
```diff
@@ -383,36 +383,47 @@ also work in retina screen::
              bundle_identifier=None
              info_plist={
-             	'NSHighResolutionCapable': 'True'
+             	'NSHighResolutionCapable': True
```
Please mind the same error in line 378

itsayellow commented:
Sure thing!  I appreciate good documentation and was about to volunteer if you didn't ask.  :)

For the edits, I tried to do as requested. (doh!  Sorry about the missing commas.)  Let me know how that looks.

htgoebel commented on doc/spec-files.rst:
```diff
@@ -370,49 +370,44 @@ of ``Info.plist``.)
+|PyInstaller| creates :file:`Info.plist` from the info_plist dict
+using the Python Standard Library module plistlib_.
+plistlib can handle nested python objects (which are translated to nested
+XML.)  Python data types are translated to the proper :file:`Info.plist`
```
"XML)." (dot behind parent)
or even
"XML), where …"

htgoebel commented on doc/spec-files.rst:
```diff
@@ -370,49 +370,44 @@ of ``Info.plist``.)
+
+Also in the above example is the key ``CFBundleDocumentTypes``.
+This entry in :file:`Info.plist` tells Mac OS X
+what filetypes your app supports.  (see `Apple document types`_).
```
dot and space after "supports".

htgoebel commented on doc/spec-files.rst:
```diff
@@ -370,49 +370,44 @@ of ``Info.plist``.)
-and follow it with a statement such as::
-
-    cp -f Info.plist dist/myscript.app/Contents/Info.plist
+For example, when you use PyQt5,
```
Is NSHighResolutionCapable related to PyQt5? If not, I'd remove the "when you use PyQt5,".

itsayellow replied:
This was existing in the documentation, and I didn't understand it either so I left it.  To tell you the truth I've never used NSHighResolutionCapable.  Usually the Info.plist key that's crucial for an application to have retina resolution enabled is `'NSPrincipleClass': 'NSApplication'`.

itsayellow replied:
https://developer.apple.com/library/content/documentation/General/Reference/InfoPlistKeyReference/Articles/CocoaKeys.html#//apple_ref/doc/uid/TP40009251-SW11
https://developer.apple.com/library/content/documentation/GraphicsAnimation/Conceptual/HighResolutionOSX/Explained/Explained.html

Based on those two links, I'm starting to think that NSHighResolutionCapable is a red herring and should be removed from the example.

It seems that the crucial thing to enable retina is to inform macOS that the application is a Cocoa app, using `'NSPrincipleClass': 'NSApplication'`.  Then, the default value of `NSHighResolutionCapable` should already be `True`

htgoebel commented on doc/spec-files.rst:
```diff
@@ -370,49 +370,44 @@ of ``Info.plist``.)
+The value of that key in :file:`Info.plist` is a list of dicts,
+each containing up to five key:value pairs.  This is represented in the
+info_plist argument by assigning to the key ``CFBundleDocumentTypes`` a value
+that is a python list containing python dicts.
```
I do not understand this two sentences. Maybe just remove them?

itsayellow replied:
Yes, I was trying to use what was already there, but quite right--this got confusing.  Removing them.

itsayellow commented:
I've fixed the punctuation goofs (and rephrased one, let me know how it looks).

What do you think about removing `NSHighResolutionCapable` from example and discussion, and instead putting `'NSPrincipleClass': 'NSApplication'`?  (See discussion above.)

htgoebel commented:
Thanks for the update. I don't like red herrings :-) 

Still I would like to have *any* other example for a boolean entry to show this has to be a boolean (and not a string or number). I would not care about using some pure example like `aBooleanValue: True`. If you know a real-life example, we could of course use it.

itsayellow commented:
I agree, although my father loved (actual) red herrings...

I updated the example with a common boolean key/value.  I added 'NSPrincipleClass'.  Finally I cleaned up the explanation of the example.

Let me know how it looks!

itsayellow commented:
BTW, I will be unplugged and away on vacation starting Saturday morning.  I'm happy if you want to make any changes in this pull request.

htgoebel commented:
Thanks for nagging :-)

Thanks you very much for this pull-request. I appreciated the discussion and really like the result. Esp. your explanation for the boolean example is even more elaborated than I your have expected it it be.

Thanks again ans I'm looking forward for more contributions from you. Maybe you are interested in working on [other OS X related issues](https://github.com/pyinstaller/pyinstaller/issues?q=is%3Aopen+is%3Aissue+label%3A%22OS+X%22). Envoy your holidays.

itsayellow commented:
Thanks for the good words, and I was happy to contribute to such a friendly project!  I will definitely take a look at the OS X issues (I already did a little bit.)
